### ISSUE TITLE
:terminal 256 colors not correct RGB values when 'termguicolors' is in effect.

### ISSUE THREAD

**Issue by bluz71** [neovim/neovim#7018]
Hello support,

`neovim` 0.2.0.1
gnome-terminal on Linux, but I also see the same issue with iTerm2 on Mac.

Basically I am not seeing the correct 256-colors in the inbuilt Neovim terminal that I see just in a plain terminal when `termguicolors` is enabled.

I expect to see [this](http://www.calmar.ws/vim/256-xterm-24bit-rgb-color-chart.html) color palette (16-255). Let's use *blue 69* as the example; it should be *#5f87ff*.

[This script](https://raw.githubusercontent.com/roosta/colors/master/scripts/support/256-colors.sh) is handy to see what is happening. Please run it as mentioned below.

Run that script just in plain terminal, then take a screenshot and use a color picker (like Gimp) to validate the color of *blue 69*; in gnome-terminal I get the correct value of *#5f87ff*.

Now do the same inside Neovim with `termguicolors` set in a `:terminal` window, run the `256-colors.sh` script, take a screenshot and do the same color picker validation. *blue 69* actually will be *#3366ff*, not what we want or expect.

I tried setting `let g:terminal_color_69 = #5f87ff`, but that had no effect. Only `g:terminal_color_0` through `g:terminal_color_15` appear to be settable.

What do I want?

1. I expect, by default, that Colors 16 - 255 in the Neovim terminal to be exactly [these values](http://www.calmar.ws/vim/256-xterm-24bit-rgb-color-chart.html)

2. I expect `g:terminal_color_16` through `g:terminal_color_255` to be settable by users just like the first 16 colors.

Item number 1. is way more important than item 2. To be honest if item 1. was fixed I could live without item 2.

Many thanks.

**Comment by justinmk** [comment]
This is discussed in closed issues, please search there. Work around is documented at `:h terminal-emulator-config`.

Nothing we can do about it.

**Comment by bluz71** [comment]
For clarity, this is a valid open issue?

**Comment by DarkDefender** [comment]
Because of a limitation in the terminal lib we use, you can only set the first 16 colors. The rest of them are hard coded sadly.

**Comment by bluz71** [comment]
@DarkDefender,

I don't really need to set the value of colors 16-255 **if** the hard-coded values that Neovim used for those colors in `:terminal` were the exact and existing *xterm 256* colors when `termguicolors` is set.

My Googling seems to indicate that 256-color terminals all use the same colors for 16-255:

[256 COLORS - CHEAT SHEET](https://jonasjacek.github.io/colors/)

[256 xterm RGB color chart](http://www.calmar.ws/vim/256-xterm-24bit-rgb-color-chart.html)

[xterm color table Vim](https://github.com/guns/xterm-color-table.vim)

Is there a reason why *CornflowerBlue* (color 69) is *#3366ff* in the Neovim terminal as compared to the expected, and desired, value of *#5f87ff* (when `termguicolors` is enabled)? That is the main question I am asking.

Basically, if the hard-coded values for 16-255 in the Neovim terminal matched standard 256-terminal-colors then I would be good. The current Neovim term colors are slightly off, hence my `$LS_COLORS` looks wrong.

Apologies if I'm repeating myself, I just wanted to emphasise that correct coloring for 16-255 is more important than user settable 16-255 colors (which appears to now be an enhancement request, though I am pretty confident it won't be needed if the coloring could be corrected).

Thanks for your patience.

**Comment by bluz71** [comment]
Answering myself, maybe *libvterm* is hard-coding 16-255 colors and maybe those hard-coded values are slightly off? *Off* compared to existing 256 terminal color values that folks like me expect?

**Comment by DarkDefender** [comment]
Here you have how libvterm generates the colors:
https://github.com/neovim/libvterm/blob/master/src/pen.c#L46

**Comment by leonerd** [comment]
Oh.. Is this all just so neovim can remember which 256-colour palette entry was used to generate the cell?

If that's all you need, I could just add that feature to remember it ;) Then you wouldn't need to fiddle with the palette.

**Comment by DarkDefender** [comment]
@leonerd Well, currently (if nothing has changed since last I did the color lookup stuff) neovim in non true color mode sets colors 1-16 to [0,0,colorindex] to figure out what pallet index to send to the user terminal.

When I first implemented it, I would have done so for all the 256 colors. However as you know this was not possible so for the remaining colors we try to figure the index out in other ways (that doesn't always work).

However, for true color mode we do not do any of this and just use the colors libvterm gives us.

And this is where @bluz71 problem comes from. It seems like the default colors for does not match what he is expecting. And he can't modify the colors manually either.

**Comment by dkarter** [comment]
Here is an example screenshot of the differences:

Left - NeoVim Terminal with `termguicolors`, Right - iTerm2
<img width="1746" alt="image" src="https://user-images.githubusercontent.com/551858/34473982-c913f8d6-ef3e-11e7-8ad3-d6954a911479.png">

**Comment by bluz71** [comment]
If it is any consolation, Vim 8's terminal color story is less configurable than Neovim. Terminal colors 1 to 16 are not settable in Vim 8, unlike Neovim, hence MacVim and gVim will have very wonky terminal colors, see [#2353](https://github.com/vim/vim/issues/2353)

So there are two sets of colors to deal with: 1 to 16 (Neovim already provides hooks to override these).
Then there are colors 17 - 256 which are hard-coded it seems by libvterm. I just wish libvterm used the standard [xterm-256](http://www.calmar.ws/vim/256-xterm-24bit-rgb-color-chart.html) hardcodes. Failing that it would be nice to provide overrides if libvterm supported it.

**Comment by leonerd** [comment]
> I just wish libvterm used the standard xterm-256 hardcodes. Failing that it would be nice to provide overrides if libvterm supported it.

I can adjust them and/or provide an interface to let them be changed. I'm quite open to feature requests :)

**Comment by bluz71** [comment]
Hello @leonard,

Thanks for being so open to ideas.

Re-read [this post of mine above](https://github.com/neovim/neovim/issues/7018#issuecomment-315662282)

I think the default colors 16-255 should exactly match *xterm 256* colors since that is what folks will expect. For instance color 69, known as CornflowerBlue, should exactly be `#5f87ff`. Note, if the exact *xterm 256* colors were used then I don't believe Neovim would need to provide a user option to set those colors since terminals like xterm and iTerm2 do not allow setting of those colors either; they are specific hard-codes. 

I assume this request right here in the Neovim issue tracker is enough? Is the Neovim libvterm the only version effected? 

Side note, VTE-based terminals (such as Gnome terminal) do appear to use the standard *xterm 256* colors for colors 16->255. It may worth having a look to see how they handle those colors, hard-coded in the VTE library or an API that Gnome Terminal uses. 

Yes: change hard-codes in Neovim libvterm to standard *xterm 256* colors.
No: user definable overrides for colors 16->255.

**Comment by bluz71** [comment]
Just checking in on this request and I notice that @justinmk references this issue in [#9393](https://github.com/neovim/neovim/issues/9393).

I don't understand what that issue means, but I am hopeful that one day this one will be fixed via those enhancements.

Cheers.

**Comment by copy** [comment]
@leonerd FYI, vim's libvterm's colour palette has been fixed in this commit: https://github.com/vim/vim/commit/a8fc0d3817f69b3599a1b56d74a2ddbd3139f40c#diff-651588ab7b51f2723734afb17d2afb02 

And, for reference, the code to generate xterm's palette can be found here: https://github.com/joejulian/xterm/blob/master/256colres.pl#L63-L88

**Comment by justinmk** [comment]
@copy thanks, see https://github.com/neovim/neovim/pull/9494

**Comment by copy** [comment]
@justinmk That's great. It would be good to additionally fix the default colour palette in libvterm. It's a simpler fix and doesn't require to specify the correct colours in their vimrc.

**Comment by bluz71** [comment]
The **default color** palette used by Neovim's libvterm should match [xterm-256color](https://jonasjacek.github.io/colors), that is for colors 16 --> 255.

I did not know that patch 8.0.1149 did that already for Vim (back in Sept 2017).

Hopefully soon fixed in Neovim as well via [#9494](https://github.com/neovim/neovim/pull/9494).

**Comment by bluz71** [comment]
> I did not know that patch 8.0.1149 did that already for Vim (back in Sept 2017).

Just tested a modern version of Vim (8.1.1700), and it does appear to use the [xterm-256colors](https://jonasjacek.github.io/colors) for colors 16 --> 255 when `termguicolors` is enabled, which is very nice. So this issue is fixed in Vim, and hopefully soon also in Neovim.

But the Vim `:terminal` itself is inferior to Neovim's, Vim's terminal buggers up my Git colors (0-15 user set colors, it defaults to bright versions for no good reason) and I have no colored Bash prompt (don't know why).

**Comment by justinmk** [comment]
#9494

**Comment by bluz71** [comment]
Just upgraded to Neovim 0.4

Unfortunately, the issue reported in the first post of this thread remains unresolved.
Fortunately, @copy appears to have a [PR](https://github.com/neovim/libvterm/pull/4) that does resolve this issue.

With that live PR on the books, this issue can remain closed, a solution is already in the queue.

Thank you @copy, hopefully merged very soon and released in a future 0.4.X release.

**Comment by bluz71** [comment]
Actually, I now request this issue be re-opened, that is until [this PR](https://github.com/neovim/libvterm/pull/4) is merged.

This issue was **not** resolved in the 0.4 release via [#9494](https://github.com/neovim/neovim/pull/9494). Hence, leaving this issue resolved is misleading. It very much still exists.

Thanks.

**Comment by bfredl** [comment]
We would like to avoid to diverge from upstream libvterm if possible. @leonerd is there a reason that https://github.com/vim/vim/commit/a8fc0d3817f69b3599a1b56d74a2ddbd3139f40c#diff-651588ab7b51f2723734afb17d2afb02 couldn't be merged into libnvim itself ?

**Comment by blueyed** [comment]
Also, isn't this related/fixed by https://github.com/neovim/neovim/pull/10994 then?

**Comment by leonerd** [comment]
I'm not sure I want to just outright change the palette like that. I'd accept an option to preselect one or the other - or perhaps extend the existing mechanism to customise the first 16 slots into customising any of the 256 slots. But a compromise of an option would be fine.

**Comment by bfredl** [comment]
@blueyed yes, but as it is a new feature (rather than a bug fix) we might not want #10994 in 0.4.x . Also it won't fix incorrect 255 colors in GUI (one would expect an embedded terminal to use the xterm standard colors also in GUI).

**Comment by bfredl** [comment]
@leonerd but what is the reason that libvterm diverges from xterm to start with?

**Comment by leonerd** [comment]
Commonality with most other terms. I had a look at a few and it turned out xterm was an odd exception and many others go for the equal-spacing ramp

**Comment by bfredl** [comment]
Ok, so correct colors are open to debate :) . #10994 should fix it for TUI (just do what the host does), while we might need an option for GUI then.

**Comment by bluz71** [comment]
> Commonality with most other terms. I had a look at a few and it turned out xterm was an odd exception and many others go for the equal-spacing ramp

That is not my experience at all.
ITerm2 and Gnome-Terminal both use the xTerm 256 colours, precisely.
It is the main reason why Vim changed libvterm’s colours to match the exact xTerm 256 colours. If not why did they change it? To me that Vim change is correct.

I dispute the assertion that xTerm’s 256 are unique to it.

**Comment by bluz71** [comment]
Note I don’t mind an option to customise the colours 16-255.
As a Vim theme author I would hard code the xTerm 256 colours.

But I believe this to be over-engineering since all the main terminals appear to use the xTerm 256 colour palette, and we can add Vim 8 to that list as well.

**Comment by bluz71** [comment]
> Also, isn't this related to / fixed by #10994 then?

This looks like it would fix the issue, yes.
I would be happy with this.
GUIs would still be an issue, but I care less about this.

**Comment by blueyed** [comment]
@bluz71 
Please try it / provide feedback there (#10994) then.
Otherwise I agree with your comments - thanks for your input.

**Comment by bluz71** [comment]
Is there a prebuilt binary I can test, a nightly of some sort?

**Comment by blueyed** [comment]
No.  Don't worry then - although building it yourself should not be hard.  See https://github.com/neovim/neovim/wiki/Building-Neovim.

**Comment by bluz71** [comment]
> I'm not sure I want to just outright change the palette like that. I'd accept an option to preselect one or the other - or perhaps extend the existing mechanism to customise the first 16 slots into customising any of the 256 slots. But a compromise of an option would be fine.

@leonerd,

Why can't the value of `$TERM` be the option that libvterm uses, without user intervention?

If I set it to `xterm-256color` or `screen-256color` then I believe libvterm should use the standard xTerm 256 colours as [listed here](https://jonasjacek.github.io/colors).

iTerm2 for Mac and Gnome-terminal on Linux both definitely use that color palette, either by default or when `$TERM` is set appropriately.

As noted already, Vim itself hard-codes the xTerm256 colors instead of using the libvterm default colors since [Sep 2017](https://github.com/vim/vim/commit/a8fc0d3817f69b3599a1b56d74a2ddbd3139f40c#diff-651588ab7b51f2723734afb17d2afb02).

**Comment by bluz71** [comment]
Merged PR [#10994] is almost perfect.

That PR solves many issues noted here except for [this new issue I just reported](https://github.com/neovim/neovim/issues/11335).

**Comment by bluz71** [comment]
For TUI Neovim #10994 does solve this issue.

In #11335 the solution ended up being to not set `g:terminal_colors` when terminal pass through is available.

For GUI Neovims the `xterm-256color` color palette will not be in effect, rather the libvterm defaults will. That's fine since I can't expect a GUI client to default to xterm 256 colors.

This issue is solved with #10994 in the upcoming 0.5.0 release.
Thanks.

**Comment by michael-angelozzi** [comment]
I think its not quite sorted, as I am using NeoVim-qt 0.5 in windows 10, and with termguicolors in affect, colors 16-255 are exact repeats of colors 0-15. The pallette is just duplicated over and over. 

Yes one can set can customize colors 0-15, to get a darker red for example, but then it messes up the terminal colors.

**Comment by bluz71** [comment]
GUI versions of Neovim don't use ‘termguicolors‘. I don't believe this is an issue with this colour scheme.

I suggest you open an issue at the Neovim-QT issue-tracker.
